snekbox/utils/timed.py

```diff
@@ -1,37 +1,34 @@
 """Calling functions with time limits."""
-import multiprocessing
-from collections.abc import Callable, Iterable, Mapping
-from typing import Any, TypeVar
+import signal
+from collections.abc import Generator
+from contextlib import contextmanager
+from typing import TypeVar
 
 _T = TypeVar("_T")
 _V = TypeVar("_V")
 
-__all__ = ("timed",)
+__all__ = ("time_limit",)
 
 
-def timed(
-    func: Callable[[_T], _V],
-    args: Iterable = (),
-    kwds: Mapping[str, Any] | None = None,
-    timeout: float | None = None,
-) -> _V:
+@contextmanager
+def time_limit(timeout: int | None = None) -> Generator[None, None, None]:
     """
-    Call a function with a time limit.
+    Decorator to call a function with a time limit. Uses SIGALRM, requires a UNIX system.
 
     Args:
-        func: Function to call.
-        args: Arguments for function.
-        kwds: Keyword arguments for function.
         timeout: Timeout limit in seconds.
 
     Raises:
         TimeoutError: If the function call takes longer than `timeout` seconds.
     """
-    if kwds is None:
-        kwds = {}
-    with multiprocessing.Pool(1) as pool:
-        result = pool.apply_async(func, args, kwds)
-        try:
-            return result.get(timeout)
-        except multiprocessing.TimeoutError as e:
-            raise TimeoutError(f"Call to {func.__name__} timed out after {timeout} seconds.") from e
+
+    def signal_handler(_signum, _frame):
+        raise TimeoutError(f"time_limit call timed out after {timeout} seconds.")
+
+    signal.signal(signal.SIGALRM, signal_handler)
+    signal.alarm(timeout)
+
+    try:
+        yield
+    finally:
+        signal.alarm(0)
```

Comment:
Clever. I don't know enough about signals to say whether this is an appropriate use, but I think it's low risk to try.

Reply:
It seems to work pretty accurately for interrupting `time.sleep` and built-in C functions in these tests [1af5bf0](https://github.com/python-discord/snekbox/pull/173/commits/1af5bf0a92a5a65346460590054c1834ff7b7ccc), so it seems okay for our usecase, and avoids the potential broken pipes issue with multiprocessing.